torchcrf/__init__.py

```diff
@@ -226,50 +232,41 @@ def decode(self, emissions: torch.Tensor,
             # shape: (batch_size, num_tags, 1)
             broadcast_score = score.unsqueeze(2)
 
-            # Broadcast emission score for every possible current tag
-            # shape: (batch_size, 1, num_tags)
-            broadcast_emission = emissions[i].unsqueeze(1)
-
             # Compute the score tensor of size (batch_size, num_tags, num_tags) where
             # for each sample, entry at row i and column j stores the score of the best
             # tag sequence so far that ends with transitioning from tag i to tag j and emitting
             # shape: (batch_size, num_tags, num_tags)
-            next_score = broadcast_score + self.transitions + broadcast_emission
+            next_score = broadcast_score + self.transitions + broadcast_emissions[i]
 
             # Find the maximum score over all possible current tag
             # shape: (batch_size, num_tags)
-            next_score, indices = next_score.max(dim=1)
+            next_score, next_indices = next_score.max(dim=1)
 
             # Set score to the next score if this timestep is valid (mask == 1)
             # and save the index that produces the next score
             # shape: (batch_size, num_tags)
             score = torch.where(mask[i].unsqueeze(1), next_score, score)
-            history.append(indices)
+            indices = torch.where(mask[i - 1].unsqueeze(1), next_indices, indices)
+            history[-i] = indices
 
         # End transition score
         # shape: (batch_size, num_tags)
         score += self.end_transitions
 
-        # Now, compute the best path for each sample
-
-        # shape: (batch_size,)
-        seq_ends = mask.long().sum(dim=0) - 1
-        best_tags_list = []
-
-        for idx in range(batch_size):
-            # Find the tag which maximizes the score at the last timestep; this is our best tag
-            # for the last timestep
-            _, best_last_tag = score[idx].max(dim=0)
-            best_tags = [best_last_tag.item()]
+        # Now, compute the best tag sequences for each sample
+        # shape: (batch_size, seq_length)
+        best_tags = emissions.new_empty(batch_size, seq_length, dtype=torch.long)
```

Comment:
Could you please use `new_zeros` here? The tag IDs must be between 0 and `num_tags` to avoid error if they're to be used by the caller.